File: frontend/src/lib/stats.js
/**
 * 実数値の即時プレビュー用ロジック。
 * バックエンド (backend/app/calculations/stats.py) と同じ計算式のJS版。
 * 保存時の正としてのバリデーション・計算はバックエンドが担うため、
 * ここではUXのための軽量な再実装として重複を許容している。
 */

export const STAT_KEYS = ['hp', 'attack', 'defense', 'special_attack', 'special_defense', 'speed'];

export const STAT_LABELS = {
	hp: 'HP',
	attack: 'こうげき',
	defense: 'ぼうぎょ',
	special_attack: 'とくこう',
	special_defense: 'とくぼう',
	speed: 'すばやさ'
};

export const NATURES = {
	hardy: [null, null],
	docile: [null, null],
	serious: [null, null],
	bashful: [null, null],
	quirky: [null, null],
	lonely: ['attack', 'defense'],
	brave: ['attack', 'speed'],
	adamant: ['attack', 'special_attack'],
	naughty: ['attack', 'special_defense'],
	bold: ['defense', 'attack'],
	relaxed: ['defense', 'speed'],
	impish: ['defense', 'special_attack'],
	lax: ['defense', 'special_defense'],
	timid: ['speed', 'attack'],
	hasty: ['speed', 'defense'],
	jolly: ['speed', 'special_attack'],
	naive: ['speed', 'special_defense'],
	modest: ['special_attack', 'attack'],
	mild: ['special_attack', 'defense'],
	quiet: ['special_attack', 'speed'],
	rash: ['special_attack', 'special_defense'],
	calm: ['special_defense', 'attack'],
	gentle: ['special_defense', 'defense'],
	sassy: ['special_defense', 'speed'],
	careful: ['special_defense', 'special_attack']
};

export const NATURE_LABELS_JA = {
	hardy: 'がんばりや',
	lonely: 'さみしがり',
	brave: 'ゆうかん',
	adamant: 'いじっぱり',
	naughty: 'やんちゃ',
	bold: 'ずぶとい',
	docile: 'すなお',
	relaxed: 'のんき',
	impish: 'わんぱく',
	lax: 'のうてんき',
	timid: 'おくびょう',
	hasty: 'せっかち',
	serious: 'まじめ',
	jolly: 'ようき',
	naive: 'むじゃき',
	modest: 'ひかえめ',
	mild: 'おっとり',
	quiet: 'れいせい',
	bashful: 'てれや',
	rash: 'うっかりや',
	calm: 'おだやか',
	gentle: 'おとなしい',
	sassy: 'なまいき',
	careful: 'しんちょう',
	quirky: 'きまぐれ'
};

export const MAX_EV_PER_STAT = 252;
export const MAX_EV_TOTAL = 510;
export const MAX_IV = 31;

export function natureModifier(nature, statKey) {
	if (statKey === 'hp') return 1;
	const pair = NATURES[nature];
	if (!pair) return 1;
	const [boosted, lowered] = pair;
	if (statKey === boosted) return 1.1;
	if (statKey === lowered) return 0.9;
	return 1;
}

export function calculateStat(base, iv, ev, level, statKey, nature) {
	const inner = Math.floor(((2 * base + iv + Math.floor(ev / 4)) * level) / 100);
	if (statKey === 'hp') {
		return inner + level + 10;
	}
	const value = inner + 5;
	return Math.floor(value * natureModifier(nature, statKey));
}

export function calculateStats(baseStats, ivs, evs, nature, level = 50) {
	const result = {};
	for (const key of STAT_KEYS) {
		result[key] = calculateStat(baseStats[key] ?? 0, ivs[key] ?? MAX_IV, evs[key] ?? 0, level, key, nature);
	}
	return result;
}

export function totalEv(evs) {
	return STAT_KEYS.reduce((sum, key) => sum + (evs[key] ?? 0), 0);
}

File: backend/app/calculations/stats.py
"""実数値(最終ステータス)の計算ロジック。

対戦シリーズで公開されている計算式(Bulbapedia等で広く解説されている
ゲームメカニクスの数式)に基づく。式自体はルール・数式であり著作物の
複製ではないため、独自実装として扱う。
"""

from __future__ import annotations

STAT_KEYS = ("hp", "attack", "defense", "special_attack", "special_defense", "speed")

# nature -> (boosted_stat, lowered_stat)。無補正の性格は (None, None)。
NATURES: dict[str, tuple[str | None, str | None]] = {
    "hardy": (None, None),
    "docile": (None, None),
    "serious": (None, None),
    "bashful": (None, None),
    "quirky": (None, None),
    "lonely": ("attack", "defense"),
    "brave": ("attack", "speed"),
    "adamant": ("attack", "special_attack"),
    "naughty": ("attack", "special_defense"),
    "bold": ("defense", "attack"),
    "relaxed": ("defense", "speed"),
    "impish": ("defense", "special_attack"),
    "lax": ("defense", "special_defense"),
    "timid": ("speed", "attack"),
    "hasty": ("speed", "defense"),
    "jolly": ("speed", "special_attack"),
    "naive": ("speed", "special_defense"),
    "modest": ("special_attack", "attack"),
    "mild": ("special_attack", "defense"),
    "quiet": ("special_attack", "speed"),
    "rash": ("special_attack", "special_defense"),
    "calm": ("special_defense", "attack"),
    "gentle": ("special_defense", "defense"),
    "sassy": ("special_defense", "speed"),
    "careful": ("special_defense", "special_attack"),
}

MAX_EV_PER_STAT = 252
MAX_EV_TOTAL = 510
MAX_IV = 31
DEFAULT_LEVEL = 50


class ValidationError(ValueError):
    pass


def validate_evs(evs: dict[str, int]) -> None:
    for key in STAT_KEYS:
        value = evs.get(key, 0)
        if value < 0 or value > MAX_EV_PER_STAT:
            raise ValidationError(f"努力値 '{key}' は0〜{MAX_EV_PER_STAT}の範囲で指定してください")
    total = sum(evs.get(key, 0) for key in STAT_KEYS)
    if total > MAX_EV_TOTAL:
        raise ValidationError(f"努力値の合計は{MAX_EV_TOTAL}以下にしてください(現在: {total})")


def validate_ivs(ivs: dict[str, int]) -> None:
    for key in STAT_KEYS:
        value = ivs.get(key, MAX_IV)
        if value < 0 or value > MAX_IV:
            raise ValidationError(f"個体値 '{key}' は0〜{MAX_IV}の範囲で指定してください")


def nature_modifier(nature: str, stat_key: str) -> float:
    if stat_key == "hp":
        return 1.0
    nature = nature.lower()
    if nature not in NATURES:
        raise ValidationError(f"不明な性格です: {nature}")
    boosted, lowered = NATURES[nature]
    if stat_key == boosted:
        return 1.1
    if stat_key == lowered:
        return 0.9
    return 1.0


def calculate_stat(
    base: int,
    iv: int,
    ev: int,
    level: int,
    stat_key: str,
    nature: str = "hardy",
) -> int:
    inner = ((2 * base + iv + ev // 4) * level) // 100
    if stat_key == "hp":
        if base == 1:  # 一部の種族(例: タマゴのみ)を考慮した特例。基本未使用。
            return 1
        return inner + level + 10
    value = inner + 5
    modifier = nature_modifier(nature, stat_key)
    return int(value * modifier)


def calculate_stats(
    base_stats: dict[str, int],
    ivs: dict[str, int],
    evs: dict[str, int],
    nature: str,
    level: int = DEFAULT_LEVEL,
) -> dict[str, int]:
    validate_evs(evs)
    validate_ivs(ivs)
    return {
        key: calculate_stat(
            base=base_stats[key],
            iv=ivs.get(key, MAX_IV),
            ev=evs.get(key, 0),
            level=level,
            stat_key=key,
            nature=nature,
        )
        for key in STAT_KEYS
    }
